### ISSUE TITLE
Changing the Icon Theme is broken (in Qt 5.4.1)

### ISSUE THREAD

**Issue by DrTobe** [lxde/libqtxdg#65]
When an lxqt application calls
```cpp
XdgIcon::fromTheme(const QString& iconName, const QIcon& fallback)
```
with a relative path, this will redirect to
```cpp
new QIcon(new QtXdg::QIconLoaderEngineFixed(name));
```
As `QIconLoader::setThemeName(...)` does not get called, `m_userTheme` is not set. Therefore, the class is looking for system themes in `systemThemeName()`. As **that** function is broken (in my case, Qt 5.4.1), the class will fall back to `fallbackTheme()` which always yields "hicolor" for me.

As all of this is happening in the `qiconloader.cpp` and this file seems to be copied from Qt, I guess that my Qt version could be responsible for this issue.

I could resolve this issue by changing the `systemThemeName()` function like that:
```cpp
static inline QString systemThemeName()
{
    if (const QPlatformTheme *theme = QGuiApplicationPrivate::platformTheme()) {
        const QVariant themeHint = theme->themeHint(QPlatformTheme::SystemIconThemeName);
        bool isEmpty = themeHint.toString().isEmpty(); // In my case, themeHint is always just an empty string: ""
        if (themeHint.isValid() && !isEmpty)
            return themeHint.toString();
    }
    return QIcon::themeName();
}
```

Finally, I am wondering why this file, which is explicitly not part of the Qt API, was copied to libqtxdg. Isn't that inherently leading to problems with future Qt versions? Could someone explain to me why we need that file and what it does for us? Why can't we just use `QIcon::setThemeName(themeName)` and `QIcon::themeName()`?

**Comment by luis-pereira** [comment]
@DrTobe Do you have lxqt-qtplugin installed ?
Sorry for the late response.

**Comment by DrTobe** [comment]
No, I haven't. Do you think it could resolve this issue?

**Comment by luis-pereira** [comment]
Yes, Pls try it.

**Comment by DrTobe** [comment]
Okay, that helped. Thank you very much.

Normally, I am not too concerned about theming and visual appearance but in this case, I was disappointed because some icons were completely missing after the installation. Could we avoid this kind of problem for other users? I have installed LXQt from source code.

**Comment by luis-pereira** [comment]
@DrTobe You have to be more specific.

**Comment by DrTobe** [comment]
I installed the panel, and some plugins had no icons. That was disappointing. I wonder if lxqt could ship with these necessary icons.

**Comment by tsujan** [comment]
> I installed the panel, and some plugins had no icons.

Which icon theme did you use?

**Comment by luis-pereira** [comment]
@DrTobe

- What plugins ?
- What Icons ?
- What Icon Theme ?

**Comment by DrTobe** [comment]
- volume plugin -> all icons missing (audio-volume-high, audio-volume-medium, audio-volume-low, audio-volume-muted)
- mount plugin -> drive-removable-media, media-eject
- lxqt-qtplugin was not installed so fallback-theme "hicolor" was used

**Comment by tsujan** [comment]
@DrTobe
There should be a reason why those icons are missing for you and not for others. lxqt-qtplugin was an essential part of LXQT, that you installed by following @luis-pereira's guidance. And you haven't told us yet which icon theme you use. A KDE-compatible icon theme works better with LXQT.

**Comment by DrTobe** [comment]
Now, with lxqt-qtplugin, I can choose different icon themes. The "gnome" theme for example has all the necessary icons.

All I wanted to say is: It did not work (icons were missing) without lxqt-qtplugin. So either the fallback-theme should contain these icons or there should be a dependency on lxqt-qtplugin.

**Comment by tsujan** [comment]
> icons were missing without lxqt-qtplugin

They should have been missing! It's a job of lxqt-qtplugin to make Qt5 apps know about icon themes.

**Comment by tsujan** [comment]
For building from the source, please read https://github.com/lxde/lxqt/wiki/Building-From-Source

**Comment by DrTobe** [comment]
Lxqt-panel does not depend on lxqt-qtplugin so you can install it without having lxqt-qtplugin installed. As a user, you just encounter that the icons are missing which can be quite frustrating. And lxqt-config does not depend on lxqt-qtplugin either, so changing the icon themes does not work properly. That's what I mean. So, I guess that lxqt-config (lxqt-config-appearance) should be dependent on lxqt-qtplugin while there should be at least a note (or warning) why icons could be missing in lxqt-panel. Otherwise, a user could install lxqt-panel and just not know why the icons don't show up.

**Comment by agaida** [comment]
it's not a dependency, given the core functionality works - this funcitionality is extended by lxqt-qtplugin - so its clearly a recommend, not a dependency

**Comment by tsujan** [comment]
@DrTobe 
Dependency can have two different meanings: real dependencies and the dependencies of downstream installable packages. When app1 doesn't depend on app2 *in its code*, it means you could compile and install app1 without having app2. However, if app2 is needed for a good UX, the packager should make "app1.deb" dependent on "app2.deb" (or .rpm, .xz, etc.).

If you take a look at the above link, you'll see that it has listed lxqt-qtplugin as a *major component*. Usually, packagers make a "meta" package, by whose installation the users will get all such components. But when someone wants to directly compile a DE from the source, the only way is consulting with such a documentation.

**Comment by DrTobe** [comment]
@agaida I think it is a dependency because you just can not change the icon theme without lxqt-qtplugin which is a core functionality of lxqt-config-appearance.

@tsujan Clearly, this problem does not arise if someone installs from their package manager but other ones build from source and they should know why the icons are missing if they haven't installed lxqt-qtplugin.

**Comment by agaida** [comment]
as @tsujan explained - this is not a build dependency, so it should be mentioned somewhere - with a big fat markup

**Comment by DrTobe** [comment]
Yes, it is not a __build__ dependency, but nevertheless, the functionality of lxqt-config-appearance depends on lxqt-qtplugin. I guess, we agree now ;). Well, I will add a note in the wiki ...

**Comment by tsujan** [comment]
@DrTobe 
I installed LXQT from its source the first time I knew about it but only after reading the above doc, which was explanatory enough, IMO. As a KDE user, I said to myself, "if these instructions work without problem, I'll start to trust LXQT." For me, everything was quite reliable since the beginning.

**Comment by DrTobe** [comment]
@tsujan I guess you have installed all the "major components". Then, everything is fine, but a major advantage of LXQt is that it is modular so I believe that it should be possible to not install everything.

**Comment by tsujan** [comment]
@DrTobe
You're right. In fact, before coming to LXQT, I was experimenting with Enlightenment for several months and used lxqt-qtplugin in it for my Qt5 apps. In such cases, it's the user's responsibility to know what he/she installs and why.

Anyway, any help to make the LXQT doc better is surely welcome!
